### PR TITLE
Add support for document rename

### PR DESCRIPTION
**Pull request by mattwar**
This change adds basic support in workspace layer for changing document names, folders, file path and source code kind.

Includes, new API's on Document and Solution for changing these properties and new API's on Workspace for handling changes to these properties during TryApplyChanges or independent changes from the host environment.

This PR does not include any changes to any specific workspaces (VS/MSBuild) to support this new kind of change (except AdHocWorkspace which supports all kinds.)

### CODE REVIEW

**Comment by mattwar** [comment]
@dotnet/roslyn-ide  please review

**Comment by tmat** [comment]
@dotnet/testimpact Can we use this to handle renames better?

**Comment by DustinCampbell** [comment]
Cool! Are you thinking of doing the same for projects?

**Comment by tannergooding** [comment]
@tmat, yes. It looks like we would no longer need to manually track the document remove/add events to detect a rename and could must track.this event instead.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 1]
```diff
@@ -795,6 +795,46 @@ internal void OnHasAllInformationChanged(ProjectId projectId, bool hasAllInforma
...
+                }
+
+                if (oldInfo.SourceCodeKind != newInfo.SourceCodeKind)
+                {
+                    newSolution = newSolution.WithDocumentSourceCodeKind(documentId, newInfo.SourceCodeKind);
+                }
+
+                SetCurrentSolution(newSolution);
+
+                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.DocumentInfoChanged, oldSolution, newSolution, documentId: documentId);
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs]: Hmm - general nitpick.  Should we avoid raising this if SetCurrentSolution ends up not actually doing anything?

**Comment by mattwar** [comment]
@dotnet-bot test windows_eta_open_prtest please

**Comment by genlu** [comment]
Any plan to add support for this `DocumentInfoChanged` event in VS workspace?

**Comment by mattwar** [comment]
@pilchie might know the answer to that.

**Pull request merged**